giskard/utils/worker_pool.py

```diff
@@ -0,0 +1,337 @@
+    end_time = time.monotonic() + timeout
+    while any([p.is_alive() for p in p_list]) and time.monotonic() < end_time:
+        sleep(0.1)
+
+
+def _stop_processes(p_list: List[Process], timeout: float = 1) -> List[Optional[int]]:
+    # Check if process is alive.
+    for p in p_list:
+        if p.is_alive():
+            # Try to terminate with SIGTERM first
+            p.terminate()
+    _wait_process_stop(p_list, timeout=timeout)
+
+    for p in p_list:
+        if p.is_alive():
+            # If still alive, kill the processes
+            p.kill()
+    _wait_process_stop(p_list, timeout=2)
+    exit_codes = [p.exitcode for p in p_list]
+    # Free all resources
+    for p in p_list:
+        p.close()
+    return exit_codes
+
+
+class GiskardFuture(Future):
+    def __init__(self) -> None:
+        super().__init__()
+        self.logs = ""
+
+
+@dataclass(frozen=True)
+class TimeoutData:
+    id: str
+    end_time: float
+
+
+@dataclass(frozen=True)
+class GiskardTask:
+    callable: Callable
+    args: Any
+    kwargs: Any
+    id: str = field(default_factory=_generate_task_id)
+
+
+@dataclass(frozen=True)
+class GiskardResult:
+    id: str
+    result: Any = None
+    exception: Any = None
+    logs: str = None
+
+
+def _process_worker(tasks_queue: SimpleQueue, tasks_results: SimpleQueue, running_process: Dict[str, str]):
+    # TODO(Bazire): For now, worker stops because the queues will be closed. Not the cleanest...
+    pid = os.getpid()
+
+    while True:
+        # Blocking accessor, will wait for a task
+        task: GiskardTask = tasks_queue.get()
+        # This is how we cleanly stop the workers
+        if task is None:
+            return
+        # Capture any log (stdout, stderr + root logger)
+        with redirect_stdout(StringIO()) as f:
+            with redirect_stderr(f):
+                handler = logging.StreamHandler(f)
+                logging.getLogger().addHandler(handler)
+                try:
+                    LOGGER.debug("Doing task %", task.id)
+                    running_process[task.id] = pid
+                    result = task.callable(*task.args, **task.kwargs)
+                    to_return = GiskardResult(id=task.id, result=result, logs=f.getvalue())
+                except BaseException as e:
+                    exception = "\n".join(traceback.format_exception(type(e), e, e.__traceback__))
+                    to_return = GiskardResult(id=task.id, exception=exception, logs=f.getvalue() + "\n" + exception)
+                finally:
+                    running_process.pop(task.id)
+                    tasks_results.put(to_return)
+                    logging.getLogger().removeHandler(handler)
+
+
+# Note: See _on_queue_feeder_error
+class PoolState(Enum):
+    STARTING = 0
+    STARTED = 1
+    BROKEN = 2
+    STOPPING = 3
+    STOPPED = 4
+
+
+FINAL_STATES = [PoolState.STOPPING, PoolState.STOPPED, PoolState.BROKEN]
+
+
+class WorkerPoolExecutor(Executor):
```

Comment:
I think we could also give pools optional names in case we decide to reuse it for other purposes. This name will be passed further to thread names as a prefix for example